sdk/core/azure-json/src/main/java/com/azure/json/JsonElement.java

```diff
@@ -0,0 +1,315 @@
+The disadvantage of extracting to an interface means the functionality is opened
+to other classes and could interfere with the current implementations.
+
+TODO: the methods defined in this class for JSON objects:
+    - addProperty
+    - setProperty
+    - getProperty
+    - removeProperty
+Could be extracted as abstract methods into a separate interface, perhaps called
+JsonObjectable or some other more appropriately named interface. The only (arguable)
+advantage of having these methods defined in this class the way they currently
+are is because they prevent any other sub class of JsonElement from having JSON
+object capabilities/functionalities. Helping isolate the functionality to JsonObject.
+The disadvantage of extracting to an interface means the functionality is opened
+to other classes and could interfere with the current implementations.
+
+TODO: the conversion methods at the bottom of this class may be unnecessary,
+especially in the way they are currently defined.
+
+
+If all three previous TODOs were to be done, then this class would simply be
+left with an abstract toString() method and the isX methods.
+
+*/
+
+/*
+SUGGESTION: Could rename class to e.g. JsonType, although low priority and hard
+to tell what would be the best name here.
+*/
+
+/**
+ * Abstract class that defines the basic, abstract methods that all valid JsonElement
+ * types need to define.
+ *
+ * Concrete sub classes of JsonElement should each define valid JSON types.
+ * Currently, the valid JSON types are: object, array, string, number, boolean,
+ * and null. These are defined by the JsonObject, JsonArray, JsonString, JsonNumber,
+ * JsonBoolean, and JsonNull classes respectively.
+ */
+public abstract class JsonElement extends StringWriter {
+    /**
+     * Default constructor.
+     * TODO: may need to remove this due to design guidelines? Unnecessary having
+     * this constructor defined in the source code if compiler is already adding
+     * this constructor implicitly when no other constructor is defined.
+     */
+    public JsonElement() {}
+
+    //------------------------------------------------------------------------//
+    //------------------------ Methods for JsonArray -------------------------//
+    // TODO: could be extracted into an interface instead of type checking in //
+    // JsonElement. Could create JsonArrayable interface or some other more   //
+    // appropriately named interface to be implemented by JsonArray.          //
+    //------------------------------------------------------------------------//
+
+    // All throw unless isArray is true.
+    /**
+     *
+     * @param element
+     * @return
+     * @throws InvalidJsonDataTypeException
+     */
+    public JsonArray addElement(JsonElement element) throws InvalidJsonDataTypeException {
+        // Case:
+        if(this.isArray()) { return (this.asArray()).addElement(element); }
+        // Case:
+        else { throw new InvalidJsonDataTypeException(); }
```

Comment:
Formatting suggestion to be applied everywhere:
```suggestion
        if (this.isArray()) { 
            return this.asArray().addElement(element); 
        } else { 
            throw new InvalidJsonDataTypeException(); 
        }
```